buildengine/build_go_test.go

```diff
@@ -245,3 +245,74 @@ func TestGoModVersion(t *testing.T) {
 	}
 	testBuild(t, bctx, fmt.Sprintf("go version %q is not recent enough for this module, needs minimum version \"9000.1.1\"", runtime.Version()[2:]), []assertion{})
 }
+
+func TestGeneratedTypeRegistry(t *testing.T) {
+	sch := &schema.Schema{
+		Modules: []*schema.Module{
+			{Name: "another", Decls: []schema.Decl{
+				&schema.Enum{
+					Name:   "TypeEnum",
+					Export: true,
+					Variants: []*schema.EnumVariant{
+						{Name: "A", Value: &schema.TypeValue{Value: &schema.Int{}}},
+						{Name: "B", Value: &schema.TypeValue{Value: &schema.String{}}},
+					},
+				},
+			}},
+		},
+	}
+	expected := `// Code generated by FTL. DO NOT EDIT.
+package main
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/TBD54566975/ftl/common/plugin"
+	"github.com/TBD54566975/ftl/go-runtime/server"
+	"github.com/TBD54566975/ftl/backend/protos/xyz/block/ftl/v1/ftlv1connect"
+	"github.com/TBD54566975/ftl/go-runtime/ftl"
+
+	"ftl/other"
+	"ftl/another"
+)
+
+func main() {
+	verbConstructor := server.NewUserVerbServer("other",
+		server.HandleCall(other.Echo),
+	)
+	ctx := context.Background()
+	typeRegistry := ftl.NewTypeRegistry()
+	typeRegistry.RegisterSumType(reflect.TypeFor[other.D](), map[string]reflect.Type{
+		"Bool": reflect.TypeFor[other.Bool](),
+		"Bytes": reflect.TypeFor[other.Bytes](),
+		"Float": reflect.TypeFor[other.Float](),
+		"Int": reflect.TypeFor[other.Int](),
+		"U": reflect.TypeFor[other.U](),
+		"List": reflect.TypeFor[other.List](),
+		"Map": reflect.TypeFor[other.Map](),
+		"String": reflect.TypeFor[other.String](),
+		"Struct": reflect.TypeFor[other.Struct](),
+	})
+	typeRegistry.RegisterSumType(reflect.TypeFor[other.SecondSumType](), map[string]reflect.Type{
+		"A": reflect.TypeFor[other.A](),
+		"B": reflect.TypeFor[other.B](),
+	})
+	typeRegistry.RegisterSumType(reflect.TypeFor[another.TypeEnum](), map[string]reflect.Type{
+		"A": reflect.TypeFor[another.A](),
+		"B": reflect.TypeFor[another.B](),
+	})
+	ctx = context.WithValue(ctx, "typeRegistry", typeRegistry)
```

Comment:
You can't use a string as a key for `context.WithValue()`:

> The provided key must be comparable and should not be of type string or any other built-in type to avoid collisions between packages using context. Users of WithValue should define their own types for keys. To avoid allocating when assigning to an interface{}, context keys often have concrete type struct{}. Alternatively, exported context key variables' static type should be a pointer or interface.

You'll need a distinct private struct type, typically with two helper functions in the same package. See other examples in the codebase for how this works.

Reply:
ah thanks! will update in my follow up